### PR TITLE
Remove implicit memory cap and add configuration for memory options

### PR DESCRIPTION
**Pull request by MLSTRM**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -107,6 +116,7 @@ function validateXML(options) {
 			// #endif
 
 			const valid = validationSucceeded(data.exitCode);
+			worker.terminate();
```
Comment (MLSTRM) [inline src/index.js]: Added this as from local testing I was able to identify that the workers weren't getting cleaned up after completion.

File: README.md [hunk 2]
```diff
@@ -59,6 +59,8 @@ async function example() {
     // All the schema files that are required to validate the documents.
     // The main XSD should be first in the array, followed by its possible dependencies.
     schema: [mySchemaFile],
+    initialMemory: 256 // Initial memory capacity in Web Assembly memory pages (1 = 6.4KiB) - 256 is minimum and default here (16MiB)
```
Comment (noppa) [inline README.md]: Maybe we should name these `initialMemoryPages` and `maxMemoryPages` or something. With the current names it's hard for a reader to know what the unit is (bytes, KiB, MiB?). 

I'm also thinking that we could also export an object with some predefined values for these for ease of use, like
```javascript
const memoryPages = {
	MiB: 16,
	GiB: 16384,
	default: 256,
	max: 65536,
};
```
so that it would be a bit more ergonomic to use like
```javascript
  initialMemoryPages: 2 * memoryPages.GiB,
```
That's probably out of scope for this PR though and can be added later.

File: src/index.js (hunk 1 above)
Comment (noppa) [reply src/index.js]: Good catch. `terminate` returns a Promise, though, so I think we should wait for that to settle too before resolving/rejecting the Promise that we return. Some tools like Jest will nag if there are dangling promises when the user's test code exits. Also maybe we should handle this in the `onerror` case too.

I think an easy way to do this would be to move the `worker` reference to outside of the `return new Promise` part, and add a `.finally` that does the cleanup.

```diff
diff --git a/src/index.js b/src/index.js
index 0f19086..ac023ca 100644
--- a/src/index.js
+++ b/src/index.js
@@ -106,6 +106,7 @@ function parseErrors(/** @type {string} */ output) {
 
 function validateXML(options) {
     const preprocessedOptions = preprocessOptions(options);
+    let worker;
 
     return new Promise(function validateXMLPromiseCb(resolve, reject) {
         function onmessage(event) {
@@ -116,7 +117,6 @@ function validateXML(options) {
             // #endif
 
             const valid = validationSucceeded(data.exitCode);
-            worker.terminate();
             if (valid === null) {
                 const err = new Error(data.stderr);
                 err.code = data.exitCode;
@@ -142,10 +142,10 @@ function validateXML(options) {
         }
 
         // #ifdef browser
-        var worker = new Worker(new URL('./xmllint-browser.mjs', import.meta.url), { type: 'module' });
+        worker = new Worker(new URL('./xmllint-browser.mjs', import.meta.url), { type: 'module' });
         // #ifdef node
         const { Worker } = require('worker_threads');
-        var worker = new Worker(require('path').resolve(__dirname, './xmllint-node.js'));
+        worker = new Worker(require('path').resolve(__dirname, './xmllint-node.js'));
         // #endif
 
         // #ifdef browser
@@ -157,7 +157,7 @@ function validateXML(options) {
         addEventListener('message', onmessage);
         addEventListener('error', onerror);
         worker.postMessage(preprocessedOptions);
-    });
+    }).finally(() => worker.terminate());
 }
 ```
Promise returned from `.finally` will be waited for before moving to the next `.then` call that the user of this library has made.

File: src/index.js [hunk 3]
```diff
@@ -38,11 +38,20 @@ function preprocessOptions(options) {
...
-	xmls.forEach(function(xml) {
-		args.push(xml['fileName']);
+	xmls.forEach(function (xml) {
+		args.push(xml[ 'fileName' ]);
 	});
 
-	return {inputFiles, args};
+	const opts = { inputFiles, args };
+
+	if (options.initialMemory) {
```
Comment (noppa) [inline src/index.js]: Nit: These conditionals are kinda unnecessary. We have the default in worker-post.js, which will work even if these come in as `undefined`, so I'd rather just write
```javascript
return {
  inputFiles, args,
  initialMemory: options.initialMemory,
  maxMemory: options.maxMemory,
}
```
I guess that sends the object keys needlessly "over the wire" to the worker if they are `undefined`, but given that we are also sending several MBs of XML, the cost of two extra property names is negligible to put it mildly.

File: src/worker-post.js [hunk 4]
```diff
@@ -29,13 +29,16 @@
...
 		// #endif
+		const wasmMemory = new WebAssembly.Memory({ initial: data.initialMemory ?? 256, maximum: data.maxMemory ?? 256 });
 
 		Module({
 			inputFiles: data.inputFiles,
 			arguments: data.args,
 			stderr: stderrBuffer.push.bind(stderrBuffer),
 			stdout: stdoutBuffer.push.bind(stdoutBuffer),
 			onExit,
+			INITIAL_MEMORY: data.initialMemory * 65536,
```
Comment (noppa) [inline src/worker-post.js]: Is this option needed even though we are providing our own `wasmMemory`? If it is, it should probably also have a default value, if `data.initialMemory` is `undefined`, this will be `NaN`.

File: src/worker-post.js [hunk 5]
```diff
@@ -29,13 +29,16 @@
 		// #ifdef node
 		var data = event;
 		// #endif
+		const wasmMemory = new WebAssembly.Memory({ initial: data.initialMemory ?? 256, maximum: data.maxMemory ?? 256 });
```
Comment (noppa) [inline src/worker-post.js]: Since we don't have Babel or other transpiling going on here, using `??` would raise the Node.js version requirement from current 12 to 14, where nullish coalescing was added. I don't have super strong desires for bending over backwards to support old Node.js versions, but I think this one is easy enough to avoid. `||` should work fine here since `0` would be quite nonsensical value for these anyway.

```suggestion
		const wasmMemory = new WebAssembly.Memory({ initial: data.initialMemory || 256, maximum: data.maxMemory || 256 });
```

File: src/index.js [hunk 6]
```diff
@@ -1,7 +1,7 @@
 'use strict';
 
 function normalizeInput(fileInput, extension) {
-	if (!Array.isArray(fileInput)) fileInput = [fileInput];
+	if (!Array.isArray(fileInput)) fileInput = [ fileInput ];
```
Comment (noppa) [inline src/index.js]: There are bunch of these unrelated formatting changes here. I don't consider these a blocker, but would be happier if they weren't here, as they make PR & commit diffs needlessly messy :upside_down_face: Maybe you have some editor setting or plugin, like Prettier plugin, that's adding these on save?

**Review by noppa — CHANGES_REQUESTED** [review]
Thanks a ton! Looks good. I did a quick test with one large XML that used to fail with out of memory error but passes nicely with this branch, so this seems to work great.

A few small code change requests, and then we'd be good to ship.

File: src/worker-post.js (hunk 4 above)
Comment (MLSTRM) [reply src/worker-post.js]: Ah this is a carry over of my attempts to get it to work with initial but not maximum, I'll remove it now as it does nothing.

File: src/index.js (hunk 6 above)
Comment (MLSTRM) [reply src/index.js]: Really having a fight with my autoformatter over this one sorry, will remove the unnecessary changes

**Comment by MLSTRM** [comment]
I've applied the suggested cleanup (hopefully not missed anything), happy to make further changes anywhere if there's more you'd like to request.

**Review by noppa — APPROVED** [review]
